model/src/main/kotlin/Provenance.kt

```diff
@@ -77,13 +77,14 @@ data class Provenance(
      * True if this [Provenance] refers to the same source code as [pkg], assuming that it belongs to the package id.
      */
     fun matches(pkg: Package): Boolean {
-        // TODO: Only comparing the hashes of the source artifacts might be sufficient.
         if (sourceArtifact != null) {
             // Note that pkg.sourceArtifact is non-nullable.
-            return sourceArtifact == pkg.sourceArtifact
+            return (sourceArtifact.hash.isNotBlank() && sourceArtifact.hash == pkg.sourceArtifact.hash) ||
```

Comment:
Thinking again, I'm not sure we should do it like this. What's not so good about this change is that the first "cheap" check ignores the `CustomData` while the latter "fallback" check *does* take `CustomData` into account.

Putting the test breakage aside (which could be adjusted), how do we want to deal with `CustomData` fields in general? Should they be considered by `equals()`? My hunch would be to say "no, they should not", but that's not easily doable with data classes. We could [omit the custom data field from the primary constructor](https://kotlinlang.org/docs/reference/data-classes.html#properties-declared-in-the-class-body), but then (de-)serialization would not work out-of-the-box.

Any opinion @mnonnenmacher, @fviernau?

Reply:
I think in this specific case the second check should only compare the `url` instead of the whole object, because if `sourceArtifact` does not have a `hash` it does not make sense to require that `pkg.sourceArtifact` also has none. But should the first check also compare the `hashAlgorithm` even if it is very unlikely that two different hash functions create the same hash?

I agree that custom data should probably not be considered in `equals` and `hashCode`, but not having it in `copy` would be bad. I think we should wait if there are ever any issue with this before we change it. And are you sure that there would be issues with deserialization? Because in general Jackson does use getters and setters and I would actually expect it to work out-of-the-box.

Reply:
> And are you sure that there would be issues with deserialization?

I haven't checked, but I thought Jackson uses the primary constructor by default, not setters in this case.

Reply:
> > I think in this specific case the second check should only compare the `url`
> 
> Is that really sufficient? Theoretically, different DNS servers could resolve the same URL to different artifacts (in an attack scenario).

But we execute the second check only if we don't have a hash, and comparing the other fields would also not help with that. Maybe we should add an option to configure if artifacts without hash should be downloaded at all.

> > And are you sure that there would be issues with deserialization?
> 
> I haven't checked, but I thought Jackson uses the primary constructor by default, not setters in this case.

I think Jackson can use a combination of constructor and setters in deserialization, but I haven't checked either.

Reply:
As discussed separately, let's postpone this discussion and for now only align on disregarding custom data when comparing provenance.